File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
import java.util.List;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        // Build the hash map for O(1) lookups
        Map<String, String> map = new HashMap<>();
        for (List<String> pair : knowledge) {
            map.put(pair.get(0), pair.get(1));
        }
        
        StringBuilder res = new StringBuilder();
        StringBuilder currKey = new StringBuilder();
        boolean inside = false;
        
        for (char c : s.toCharArray()) {
            if (c == '(') {
                inside = true;
            } else if (c == ')') {
                inside = false;
                String key = currKey.toString();
                // Append the mapped value or "?" if it doesn't exist
                res.append(map.getOrDefault(key, "?"));
                currKey.setLength(0); // Efficiently clear the current key
            } else if (inside) {
                currKey.append(c);
            } else {
                res.append(c);
            }
        }
        
        return res.toString();
    }
}

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.py
from typing import List

class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        # Convert knowledge list to a dictionary for O(1) lookups
        know_dict = {k: v for k, v in knowledge}
        
        res = []
        curr_key = []
        inside = False
        
        for char in s:
            if char == '(':
                inside = True
            elif char == ')':
                inside = False
                # Reconstruct the key string and look it up
                key_str = "".join(curr_key)
                res.append(know_dict.get(key_str, "?"))
                # Reset the temporary key container
                curr_key = []
            elif inside:
                curr_key.append(char)
            else:
                res.append(char)
                
        # Join the list into the final string
        return "".join(res)